modules/fx/arrow-fx/src/main/kotlin/arrow/fx/extensions/io.kt

```diff
@@ -43,127 +48,156 @@ import arrow.typeclasses.MonadThrow
+interface IOMonadError<E> : MonadError<IOPartialOf<E>, Throwable>, IOApplicativeError<E>, IOMonad<E> {
 
   override fun <A> just(a: A): IO<Nothing, A> = IO.just(a)
 
-  override fun <A, B> IOOf<A>.ap(ff: IOOf<(A) -> B>): IO<Nothing, B> =
-    fix().ap(ff)
+  override fun <A, B> IOOf<E, A>.ap(ff: IOOf<E, (A) -> B>): IO<E, B> =
+    Ap(ff)
 
-  override fun <A, B> IOOf<A>.map(f: (A) -> B): IO<Nothing, B> =
+  override fun <A, B> IOOf<E, A>.map(f: (A) -> B): IO<E, B> =
     fix().map(f)
 
-  override fun <A> IOOf<A>.attempt(): IO<Nothing, Either<Throwable, A>> =
+  override fun <A> IOOf<E, A>.attempt(): IO<E, Either<Throwable, A>> =
     fix().attempt()
 
-  override fun <A> IOOf<A>.handleErrorWith(f: (Throwable) -> IOOf<A>): IO<Nothing, A> =
-    HandleErrorWith(f)
+  override fun <A> IOOf<E, A>.handleErrorWith(f: (Throwable) -> IOOf<E, A>): IO<E, A> =
+    HandleErrorWith(f, { e -> IO.raiseError(e) })
 
-  override fun <A, B> IOOf<A>.redeemWith(fe: (Throwable) -> IOOf<B>, fb: (A) -> IOOf<B>): IO<Nothing, B> =
-    fix().redeemWith(fe, fb)
+  override fun <A, B> IOOf<E, A>.redeemWith(fe: (Throwable) -> IOOf<E, B>, fb: (A) -> IOOf<E, B>): IO<E, B> =
+    RedeemWith({ t -> fe(t) }, { e -> IO.raiseError(e) }, { a -> fb(a) })
 
   override fun <A> raiseError(e: Throwable): IO<Nothing, A> =
     IO.raiseException(e)
 }
 
 @extension
-interface IOMonadThrow : MonadThrow<ForIO>, IOMonadError
+interface IOMonadThrow<E> : MonadThrow<IOPartialOf<E>>, IOMonadError<E>
 
 @extension
-interface IOBracket : Bracket<ForIO, Throwable>, IOMonadThrow {
-  override fun <A, B> IOOf<A>.bracketCase(release: (A, ExitCase<Throwable>) -> IOOf<Unit>, use: (A) -> IOOf<B>): IO<Nothing, B> =
-    fix().bracketCase(release, use)
-
-  override fun <A, B> IOOf<A>.bracket(release: (A) -> IOOf<Unit>, use: (A) -> IOOf<B>): IO<Nothing, B> =
-    fix().bracket(release, use)
-
-  override fun <A> IOOf<A>.guarantee(finalizer: IOOf<Unit>): IO<Nothing, A> =
-    fix().guarantee(finalizer)
-
-  override fun <A> IOOf<A>.guaranteeCase(finalizer: (ExitCase<Throwable>) -> IOOf<Unit>): IO<Nothing, A> =
-    GuaranteeCase(finalizer)
+interface IOBracket<E> : Bracket<IOPartialOf<E>, Throwable>, IOMonadThrow<E> {
+  override fun <A, B> IOOf<E, A>.bracketCase(release: (A, ExitCase<Throwable>) -> IOOf<E, Unit>, use: (A) -> IOOf<E, B>): IO<E, B> =
+    // Capture `E` into `Either`
+    RedeemWith({ t -> IO.raiseException<Either<E, A>>(t) }, { e -> IO.just(Left(e)) }, { a -> IO.just(Right(a)) })
+      .BracketCase<Either<E, A>, E, Either<E, B>>(release = { a, ex ->
+        when (a) {
+          is Either.Right -> { // Release resource.
+            when (ex) {
+              ExitCase2.Completed -> release(a.b, ExitCase.Completed).fix()
+              ExitCase2.Canceled -> release(a.b, ExitCase.Canceled).fix()
+              is ExitCase2.Exception -> release(a.b, ExitCase.Error(ex.exception)).fix()
+              is ExitCase2.Error -> throw AssertionError("Unreachable") // E is `Nothing`.
```

Comment:
Maybe worth setting the original exception as the origin of this one.

Reply:
This is not an exception but an `Error` of type `Nothing`.

Reply:
What's the type in the field in ExitCase2.Error then?

Reply:
`Nothing` so can also return `ex.error` which is `Nothing` so no need to throw.